### PR TITLE
Remove some dead code in System.Private.DataContractSerialization

### PR DESCRIPTION
**Pull request by stephentoub**
Found by a coverity code scan
cc: @shmao

### CODE REVIEW

File: src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ClassDataContract.cs [hunk 1]
```diff
@@ -1121,18 +1121,16 @@ private void ImportDataMembers()
                             DataMember memberContract = new DataMember(member);
 
                             memberContract.Name = DataContract.EncodeLocalName(member.Name);
-                            object[] optionalFields = null;
-                            if (optionalFields == null || optionalFields.Length == 0)
```
Comment (stephentoub) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ClassDataContract.cs]: optionalFields is obviously null here, so we'll always enter the body of the if

File: src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/Json/XmlJsonReader.cs [hunk 2]
```diff
@@ -916,10 +916,6 @@ private static int BreakText(byte[] buffer, int offset, int length)
                 {
                     return originalLength; // sequence fits exactly
                 }
-                if (length == 0)
```
Comment (stephentoub) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/Json/XmlJsonReader.cs]: length is already checked earlier to be non-zero

File: src/System.Private.DataContractSerialization/src/System/Xml/XmlUTF8TextReader.cs [hunk 3]
```diff
@@ -1240,8 +1240,6 @@ private int BreakText(byte[] buffer, int offset, int length)
                 }
                 if (length + byteCount == originalLength)
                     return originalLength; // sequence fits exactly
-                if (length == 0)
```
Comment (stephentoub) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlUTF8TextReader.cs]: length is already checked earlier to be non-zero

File: src/System.Private.DataContractSerialization/src/System/Xml/XmlConverter.cs [hunk 4]
```diff
@@ -555,8 +555,6 @@ private static bool TryParseInt64(byte[] chars, int offset, int count, out long
                 int offsetMax = offset + count;
                 if (chars[offset] == '-')
                 {
-                    if (count == 1)
```
Comment (stephentoub) [inline src/System.Private.DataContractSerialization/src/System/Xml/XmlConverter.cs]: count is already checked to be >= 11

File: src/System.Private.DataContractSerialization/src/System/Xml/ValueHandle.cs [hunk 5]
```diff
@@ -475,11 +475,7 @@ public byte[] ToByteArray()
 
         public string GetString()
         {
-            ValueHandleType type = _type;
-            if (type == ValueHandleType.UTF8)
```
Comment (stephentoub) [inline src/System.Private.DataContractSerialization/src/System/Xml/ValueHandle.cs]: UTF8 is already in the switch below with the same outcome

File: src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ClassDataContract.cs (hunk 1 above)
Comment (shmao) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ClassDataContract.cs]: On desktop, this is set as (see the code [here](http://referencesource.microsoft.com/#System.Runtime.Serialization/System/Runtime/Serialization/ClassDataContract.cs,981)),
```c#
object[] optionalFields = field.GetCustomAttributes(Globals.TypeOfOptionalFieldAttribute, false);
```
As we are going to add `OptionalFieldAttribute` in Core, I think we should keep the cod here.

File: src/System.Private.DataContractSerialization/src/System/Xml/ValueHandle.cs (hunk 5 above)
Comment (shmao) [inline src/System.Private.DataContractSerialization/src/System/Xml/ValueHandle.cs]: This code might be written for perf optimization. We'd better not remove this without perf measurement.

/cc: @mconnew

Comment (stephentoub) [inline src/System.Private.DataContractSerialization/src/System/Xml/ValueHandle.cs]: Then should it be removed from the switch? It doesn't need to be in both places.

Comment (shmao) [inline src/System.Private.DataContractSerialization/src/System/Xml/ValueHandle.cs]: Yes, I agree. The one in switch can/should be removed.

File: src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ClassDataContract.cs (hunk 1 above)
Comment (stephentoub) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ClassDataContract.cs]: > As we are going to add OptionalFieldAttribute in Core

Yes, it's been added here:
https://github.com/dotnet/corefx/blob/dev/api/src/System.Runtime/ref/System.Runtime.cs#L5469
along with the rest of the runtime serialization primitives.  Do you have an issue open on GitHub to add back such pieces of code that were removed?

Comment (shmao) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ClassDataContract.cs]: Opened https://github.com/dotnet/corefx/issues/11477 for tracking adding the implementation back.

Comment (stephentoub) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ClassDataContract.cs]: Updated

File: src/System.Private.DataContractSerialization/src/System/Xml/ValueHandle.cs (hunk 5 above)
Comment (stephentoub) [inline src/System.Private.DataContractSerialization/src/System/Xml/ValueHandle.cs]: Updated

**Pull request merged**